gcc/d/d-codegen.cc

```diff
@@ -49,6 +49,226 @@
+
+  // In each subblock, record that this is its superior.
+  for (tree t = level->blocks; t; t = BLOCK_CHAIN (t))
+    BLOCK_SUPERCONTEXT (t) = block;
+
+  if (level->kind == level_function)
+    {
+      // Dispose of the block that we just made inside some higher level.
+      DECL_INITIAL (current_function_decl) = block;
+      BLOCK_SUPERCONTEXT (block) = current_function_decl;
+
+      // Pop all the labels declared in the function.
+      if (cfun->language->labels)
+	cfun->language->labels->traverse<tree, &pop_label>(block);
+    }
+  else
+    {
+      // Any uses of undefined labels, and any defined labels, now operate
+      // under constraints of next binding contour.
+      if (cfun && cfun->language->labels)
+	{
+	  cfun->language->labels->traverse<binding_level *, &pop_binding_label>
+	    (level);
+	}
+
+      current_binding_level->blocks
+	= block_chainon(current_binding_level->blocks, block);
+    }
+
+  TREE_USED (block) = 1;
+  return block;
+}
+
+
+// Create an empty statement tree rooted at T.
+void
+push_stmt_list()
+{
+  tree t = alloc_stmt_list();
+  vec_safe_push(cfun->language->stmt_list, t);
+  d_keep(t);
+}
+
+// Finish the statement tree rooted at T.
+tree
+pop_stmt_list()
+{
+  tree t = cfun->language->stmt_list->pop();
+
+  // If the statement list is completely empty, just return it.  This is
+  // just as good small as build_empty_stmt, with the advantage that
+  // statement lists are merged when they appended to one another.
+  // So using the STATEMENT_LIST avoids pathological buildup of EMPTY_STMT_P
+  // statements.
+  if (TREE_SIDE_EFFECTS (t))
+    {
+      // If the statement list contained exactly one statement, then
+      // extract it immediately.
+      tree_stmt_iterator i = tsi_start (t);
+
+      if (tsi_one_before_end_p (i))
+	{
+	  tree u = tsi_stmt (i);
+	  tsi_delink (&i);
+	  free_stmt_list (t);
+	  t = u;
+	}
+    }
+
+  return t;
+}
+
+// T is an expression statement.  Add it to the statement-tree.
+
+void
+add_stmt(tree t)
+{
+  // Ignore (void) 0; expression statements received from the frontend.
+  // Likewise void_node is used when contracts become nops in release code.
+  if (t == void_node || IS_EMPTY_STMT (t))
+    return;
+
+  if (EXPR_P (t) && !EXPR_HAS_LOCATION (t))
+    {
+      if (!EXPR_HAS_LOCATION (t))
+	SET_EXPR_LOCATION (t, input_location);
+    }
+
+  tree stmt_list = cfun->language->stmt_list->last();
+  append_to_statement_list_force(t, &stmt_list);
+}
+
+//
+void
+start_function(FuncDeclaration *decl)
+{
+  cfun->language = ggc_cleared_alloc<language_function>();
+  cfun->language->function = decl;
+
+  // Default chain value is 'null' unless parent found.
+  cfun->language->static_chain = null_pointer_node;
+
+  // Find module m for this function
+  for (Dsymbol *dsym = decl->parent; dsym != NULL; dsym = dsym->parent)
+    {
+      if (dsym->isModule())
+       {
+	 cfun->language->module = (Module *) dsym;
+	 break;
+       }
+    }
+
```

Comment:
I guess every FuncDeclaration should have a valid Module so  maybe add an assert?